fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+                this.sizeStatisticsResults.maxKeySize = statsContinuation.getPartialResults().getMaxKeySize();
+                this.sizeStatisticsResults.valueSize = statsContinuation.getPartialResults().getValueSize();
+                this.sizeStatisticsResults.maxValueSize = statsContinuation.getPartialResults().getMaxValueSize();
+                List<Long> bucketList = statsContinuation.getPartialResults().getSizeBucketsList();
+                Arrays.setAll(this.sizeStatisticsResults.sizeBuckets, i -> bucketList.get(i));
+                kvCursorContinuation = statsContinuation.getContinuation().toByteArray(); //underlying KV cursor continues here
+
+            } catch (InvalidProtocolBufferException ex) {
+                throw new RecordCoreException("Error parsing SizeStatisticsCollectorCursor continuation", ex)
+                        .addLogInfo("raw_bytes", ByteArrayUtil2.loggable(continuation));
+            }
+        }
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<SizeStatisticsResults>> onNext() {
+        //Each time this is called it instantiates the underlying KV cursor, aggregates as much of its input as it can provide, and closes it
+        //Either the fully aggregated result is returned, or no value and a continuation comprised of partially aggregated results is returned
+        if (finalResultsEmitted) {
+            return CompletableFuture.completedFuture(RecordCursorResult.exhausted());
+        }
+        return subspaceProvider.getSubspaceAsync(context).thenCompose(subspace -> {
+            KeyValueCursor kvCursor = KeyValueCursor.Builder.withSubspace(subspace).setContext(this.context).setContinuation(this.kvCursorContinuation).setScanProperties(scanProperties).build();
+            return kvCursor.forEachResult(nextKv -> {
+                updateStatistics(nextKv.get());
+            }).handle((resultKv, err) -> {
+                if (err == null) {
+                    //resultKV.hasNext() is false and so determine whether this is because of in-band reason or a continuable out-of-band reason
+                    if (resultKv.getNoNextReason() == NoNextReason.SOURCE_EXHAUSTED) {
+                        finalResultsEmitted = true;
+                        this.nextStatsResult = RecordCursorResult.withNextValue(emitFinalStats(), new SizeStatisticsCollectorCursorContinuation(resultKv));
+                    } else {
+                        //the underlying cursor did not produce a row but there are more, return a continuation and propagate the underlying no next reason
+                        this.nextStatsResult = RecordCursorResult.withoutNextValue(new SizeStatisticsCollectorCursorContinuation(resultKv), resultKv.getNoNextReason());
+                    }
+                    return this.nextStatsResult;
+                } else {
+                    throw new RuntimeException(err);  //might be retriable, caller will decide what to do
```

Comment:
I think instead of instantiating a new `RuntimeException`, one wants to call the context's `asyncToSyncExceptionMapper`. Or, actually, because it's just rethrowing the error, calling `thenApply(resultKv)` instead of `handle` might be the move.

The original intention behind stopping on retriable errors was to ensure that the user wouldn't do something like scan a big ol' range and then hit some timeout halfway through and lose all progress (and also to avoid double counting ranges in the face of errors). I think this API makes it harder to accomplish that.